in-flight reauthorization Fix #18

dknoma opened the pull request:
Fix #18

dknoma committed:
creating new scripts for in-flight reauthorization: same auth, different expiration. for draft pr

jfallows commented on src/test/java/org/reaktivity/nukleus/oauth/internal/streams/StreamsIT.java:
```diff
@@ -226,6 +226,17 @@ public void shouldForwardRequestWithSetScopesIssuerAndAudienceWithValidJwtRS256O
+    })
+    public void shouldReauthorizeInFlightRequest() throws Exception
+    {
+        k3po.finish();
```
Probably need the following:
```
k3po.start();
Thread.sleep(...); // let first token expire
k3po.notifyBarrier("BARRIER_NAME");
k3po.finish();
```

dknoma committed:
Merge branch 'develop' of git://github.com/reaktivity/nukleus-oauth.java into develop

# Conflicts:
#	pom.xml

dknoma committed:
updated in-flight reauthorization test and pom.xml.

dknoma committed:
testing in-flight requests

dknoma committed:
updated same auth, update expiration test description of `Thread.sleep(...)`. First token would expire after this amount of time if its expiration is not properly updated

dknoma committed:
updated `StreamsIT` reaktor counter capacity to 8192 to handle multiple routes in a single script. added a test for rejecting in flight authorization due to insufficient privileges on the fresher incoming token.

dknoma committed:
new test for reauthorization when an incoming token has more privileges. would pass once feature is actually implemented

dknoma committed:
Merge branch 'develop' of git://github.com/reaktivity/nukleus-oauth.java into develop

# Conflicts:
#	pom.xml

dknoma committed:
added new spec scripts. implementing capture of in-flight requests and responses

dknoma committed:
committing initial version in-flight request capture.

dknoma committed:
Merge branch 'develop' of git://github.com/reaktivity/nukleus-oauth.java into develop

dknoma committed:
Initial implementation of in-flight reauthorization.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -92,6 +97,8 @@
 
     private final JsonWebSignature signature = new JsonWebSignature();
 
+    private final Map<Long, Map<String, AuthorizationState>>[] inFlightAuthorizationsByRealm = new HashMap[16];
```
Use Long2ObjectHashMap instead to avoid boxing long primitives during lookup.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -188,6 +199,46 @@ private MessageConsumer newInitialStream(
+                        final Map<String, AuthorizationState> authStateMap = inFlightAuthorizationsByRealm[realmId]
+                                .computeIfAbsent(begin.affinity(), g -> new IdentityHashMap<>());
+                        final AuthorizationState authState = authStateMap.get(subject.intern());
+                        if(authState != null)
```
Move above code to supplyAuthState method, always returning non-null AuthorizationState.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -188,6 +199,46 @@ private MessageConsumer newInitialStream(
+
+                            authState.incrementReferenceCount();
+
+                            if((authStateAuthorization & connectAuthorization) >= authStateAuthorization &&
```
Should == authStateAuthorization, not >= right?

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -188,6 +199,46 @@ private MessageConsumer newInitialStream(
+                        {
+                            final AuthorizationState newAuthState = new AuthorizationState(connectAuthorization, expiresAtMillis);
+                            newAuthState.incrementReferenceCount();
+                            authStateMap.put(subject.intern(), newAuthState);
```
Nice job using intern 👍

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -270,6 +321,59 @@ private static long expiresAtMillis(
         return expiresAtMillis;
     }
 
```
Consider renaming to OAuthAccessGrant

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -443,26 +547,58 @@ private void onSignal(
+                                                                              .orElse(null);
+                    if(authStateMap != null)
+                    {
+                        final AuthorizationState authState = authStateMap.get(subject.intern());
```
Move above methods to lookupAuthState method instead, may return null.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -443,26 +547,58 @@ private void onSignal(
+                        extendedExpiration = expiresAtMillis > System.currentTimeMillis();
+                        if(extendedExpiration)
+                        {
+                            final long delay = expiresAtMillis - System.currentTimeMillis();
```
Calculate delay first, then test for >= 0 so currentTimeMillis only called once

dknoma committed:
moved auth state modifications to a separate method to make method less bulky. also changed hashmaps to Long2ObjectHashMap

dknoma committed:
new `tryExtendAuthStateExpiration` to make `onSignal` look less bulky.

dknoma committed:
fixed some redundant variables

dknoma committed:
uncommented decrement reference count code. May want to move decrement code to some other onFrame method where the replies are actually successfully sent.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -92,6 +96,8 @@
 
     private final JsonWebSignature signature = new JsonWebSignature();
 
+    private final Map<Long, Map<String, OAuthAccessGrant>>[] inFlightAuthorizationsByRealm = new Long2ObjectHashMap[16];
```
Change the type on the left to `Long2ObjectHashMap`, otherwise it will continue to call the boxed `Long` methods instead of the primitive `long` methods for `get` and `put` through the generic `Map` interface even though the implementation is `Long2ObjectHashMap`.

Also, consider renaming `inFlightAuthorizationsByRealm` to something like `grantsBySubjectByAffinityPerRealm`.
Then, an individual entry in the array is `grantsBySubjectByAffinity`, a lookup by `BEGIN` `affinity` gives `grantsBySubject`, and the final `OAuthAccessGrant` is `grant`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -441,7 +582,8 @@ private void onSignal(
 
-            if (signalId == TOKEN_EXPIRED_SIGNAL)
+            if (signalId == TOKEN_EXPIRED_SIGNAL &&
+                !tryExtendAuthStateExpiration((int) ((sourceAuthorization & REALM_MASK) >> 48)))
```
Refer to an existing constant for `SCOPE_BITS` instead of `48`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -462,8 +604,50 @@ private void onSignal(
+
+                if(authStateMap != null)
+                {
+                    final OAuthAccessGrant authState = authStateMap.get(subject.intern());
```
Recommend moving this into `lookup` and returning `null` if `authStateMap` is `null`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -462,8 +604,50 @@ private void onSignal(
+            return inFlightAuthorizationsByRealm[sourceRealmId].values()
+                                                               .stream()
+                                                               .filter(map -> map.containsKey(subject.intern()))
+                                                               .findFirst()
```
Can this be replaced by `return inFlightAuthorizationsByRealm[sourceRealmId].get(subject.intern());` ?

dknoma replied:
That would have to be `inFlightAuthorizationsByRealm[sourceRealmId].get(affinity).get(subject.intern());` which we wouldn't have unless we store affinity as a private variable or something.

jfallows replied:
Wouldn't it just be another argument to the method, called from a context where we receive the `BEGIN` (which has `affinity` on the frame)?

dknoma committed:
made OAuthAccessGrant datastructure, variable, and method names to be more clear and consistent. updated Map to Long2ObjectMap to prevent boxed Long methods rather than primitive long methods.

dknoma committed:
now retains a reference to this streams `OAuthAccessGrant` for its frames to have access to when needing to check the grants state.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -331,23 +305,22 @@ private void modifyAuthorizationState(
 
-    private OAuthAccessGrant supplyAuthState(
-        Map<String, OAuthAccessGrant> authStateMap,
+    private OAuthAccessGrant supplyGrant(
```
Suggest the following:
```
private OAuthGrant supplyGrant(
    final realmIndex,
    final affinityId,
    final String subject)
{
    final Long2ObjectHashMap<Map<String, OAuthAccessGrant>> grantsBySubjectByAffinity = grantsBySubjectByAffinityPerRealm[realmIndex];
    final Map<String, OAuthAccessGrant> grantsBySubject = grantsBySubjectByAffinity.computeIfAbsent(affinityId, a -> new IdentityHashMap());

    final String subjectKey = subject.intern();
    final Consumer<String> cleaner = grantsBySubject::remove;
    return grantsBySubject.computeIfAbsent(subjectKey, s -> new OAuthAccessGrant(cleaner);
}
```

dknoma committed:
made suggested changes. `OAuthAccessGrant` now holds a hook to clean itself up from the datastructure, variables are modified via direct field access, and grants are still held by each `OAuthProxy` but now increment its grants reference count in its constructor. Subject is now referenced in `OAuthAccessGrant` for better debugging

dknoma committed:
fixed where decrements are happening and put asserts to make sure that cleanup is actually possible

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -130,6 +140,10 @@ public OAuthProxyFactory(
+        for(int i = 0; i < grantsBySubjectByAffinityPerRealm.length; i++)
+        {
+            grantsBySubjectByAffinityPerRealm[i] = new Long2ObjectHashMap<>();
+        }
```
Suggest using `Arrays.setAll` instead of explicit `for` loop.

```
Arrays.setAll(grantsBySubjectByAffinityPerRealm, i -> new Long2ObjectHashMap());
```

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -245,6 +261,64 @@ private RouteFW wrapRoute(
+                        grant.subject = subject;
+                        grant.authorization = connectAuthorization;
+                        grant.expiresAt = expiresAtMillis;
+                    }
```
Move this `if-else` block to a method on `OAuthAccessGrant`, perhaps called `reauthorize` returning `boolean` if successfully reauthorized according to enforced compatibility conditions.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -315,6 +414,13 @@ private OAuthProxy(
+            {
+                assert (grant.cleaner != null);
+                grant.referenceCount++;
+            }
```
Move `if` block to `aquire()` method on `OAuthAccessGrant`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -315,6 +414,13 @@ private OAuthProxy(
             }
+
+            this.grant = grant;
+            if(this.grant != null)
```
Why would `grant` be `null`, must it not always be `non-null`?
This would be preferable to limit the number of special cases.
If `subject` is `null` in `BEGIN`, perhaps we always need to create a new `OAuthAccessGrant` ?

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -462,6 +570,40 @@ private void onSignal(
+                {
+                    grant.cleaner.accept(grant.subject);
+                    grant.cleaner = null;
+                }
```
Move to `release` method on `OAuthAccessGrant`.

dknoma committed:
made requested changes; more cleaning up

dknoma committed:
added 2 more tests:
 - upgraded stream do not reauthorize -> downgraded stream (back to same roles) and reauthorize
 - downgraded stream do not reauthorize -> upgraded stream (back to same roles) and reauthorize

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -92,6 +101,9 @@
 
     private final JsonWebSignature signature = new JsonWebSignature();
 
+    private final Long2ObjectHashMap<Map<String, OAuthAccessGrant>>[] grantsBySubjectByAffinityPerRealm =
```
Assign non-static fields in constructor.
`JsonWebSignature` is pseudo-static, acting as a flyweight.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -188,13 +201,15 @@ private MessageConsumer newInitialStream(
             long connectReplyId = supplyReplyId.applyAsLong(connectInitialId);
             long expiresAtMillis = config.expireInFlightRequests() ? expiresAtMillis(verified) : EXPIRES_NEVER;
 
+            final OAuthAccessGrant grant = modifyGrant(verified, begin.affinity(), connectAuthorization, expiresAtMillis);
```
Hoist `begin.affinity()` as local variable.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -245,6 +260,58 @@ private RouteFW wrapRoute(
+        catch (InvalidJwtException | JoseException | MalformedClaimException e)
+        {
+            // TODO: diagnostics?
+        }
```
Suggest moving the above lines to `String resolveSubject(JsonWebSignature)` returning the `subject`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -245,6 +260,58 @@ private RouteFW wrapRoute(
+
+        final OAuthAccessGrant grant = supplyGrant(realmId, affinity, subject);
+        grant.reauthorize(subject, connectAuthorization, expiresAtMillis);
+        return grant;
```
Suggest inlining the remaining lines to `newInitialStream`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -270,6 +337,69 @@ private static long expiresAtMillis(
+            long expiresAtMillis)
+        {
+            final boolean reauthorized;
+            if(referenceCount > 0)
```
Missing whitespace before open paren.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -270,6 +337,69 @@ private static long expiresAtMillis(
+
+        private void release()
+        {
+            if(subject != null)
```
Missing whitespace before open paren.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -270,6 +337,69 @@ private static long expiresAtMillis(
+                cleaner.accept(subject.intern());
+            }
+            cleaner = null;
+        }
```
Where is `referenceCount--` ?

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -315,6 +448,9 @@ private OAuthProxy(
             }
+
+            this.grant = Objects.requireNonNull(grant);
+            grant.acquire();
```
Let's move this into `newInitialStream` as it has a side-effect.
Note: we'll need to call `grant.acquire()` once for initial, and once for reply as part of setting up reply throttle in advance of `newReplyStream`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -462,7 +600,31 @@ private void onSignal(
+            if (grant.referenceCount == 0)
+            {
+                grant.release();
             }
```
Move this logic inside `release()` method instead.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -483,8 +645,18 @@ private void cancelTimerIfNecessary()
             {
                 expiryFuture.cancel(true);
                 expiryFuture = null;
+                decrementGrantReferenceCount();
```
Use `grant.release()` instead.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -462,7 +600,31 @@ private void onSignal(
                 {
                     writer.doAbort(target, targetRouteId, targetStreamId, traceId, targetAuthorization);
                 }
+            }
```
Replace this with a switch on `signalId` and a `case TOKEN_EXPIRED_SIGNAL:` that delegates to `onTokenExpiredSignal(SignalFW)`.

Then in `onTokenExpiredSignal(SignalFW)` put the body of the if block above, and inline the body of `tryExtendGrantExpiration()` too.

dknoma committed:
made requested changes. removed unneeded tests. inlined code and condensed code where methods were unneeded.

dknoma committed:
updated `TOKEN_EXPIRED_SIGNAL` to be of type `int`, and changed the if statement in `onSignal` to a switch/case

dknoma committed:
removed unnecessary test

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -245,6 +267,51 @@ private RouteFW wrapRoute(
+        final Long2ObjectHashMap<Map<String, OAuthAccessGrant>> grantsBySubjectByAffinity =
+                grantsBySubjectByAffinityPerRealm[realmIndex];
+        final Map<String, OAuthAccessGrant> grantsBySubject =
+                grantsBySubjectByAffinity.computeIfAbsent(affinityId, a -> new IdentityHashMap<>());
```
Move these lookups inside `subject != null` clause.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -270,6 +337,74 @@ private static long expiresAtMillis(
+        private long expiresAt;
+        private int referenceCount;
+        private Consumer<String> cleaner;
+
```
Add no-args constructor which no-op cleaner.
```
private OAuthAccessGrant()
{
    this.cleaner = NOOP_CLEANER;
}
```

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -245,6 +267,51 @@ private RouteFW wrapRoute(
+        }
+        else
+        {
+            return new OAuthAccessGrant(this::noOp);
```
Use no-args `OAuthAccessGrant` constructor and remove now unused `noOp` method.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -270,6 +337,74 @@ private static long expiresAtMillis(
+            {
+                if (subject != null)
+                {
+                    cleaner.accept(subject.intern());
```
No need to re-intern `subject` since it was already interned before assignment, right?

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -65,7 +70,11 @@
+
+    private static final long REALM_MASK = 0xFFFF_000000000000L;
+
+    private static final int SCOPE_BITS = 48;
```
Create no-op cleaner constant.
```
private static final Consumer<String> NOOP_CLEANER = s -> {};
```

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -441,7 +582,27 @@ private void onSignal(
+        {
+            final long delay = grant.expiresAt - System.currentTimeMillis();
+
+            if (delay >= 0)
```
Strictly speaking, if `delay == 0` then token has expired, right?

dknoma committed:
applied changes. noOp -> no op Consumer constant. cleaned up minor inefficiencies and bugs.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -245,6 +269,47 @@ private RouteFW wrapRoute(
+        }
+        catch (InvalidJwtException | JoseException | MalformedClaimException e)
+        {
+            // TODO: diagnostics?
```
Change comment to `// invalid token`

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -245,6 +269,47 @@ private RouteFW wrapRoute(
+        final long affinityId,
+        final String subject)
+    {
+
```
Remove empty newline

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -245,6 +269,47 @@ private RouteFW wrapRoute(
+            final Long2ObjectHashMap<Map<String, OAuthAccessGrant>> grantsBySubjectByAffinity =
+                    grantsBySubjectByAffinityPerRealm[realmIndex];
+            final Map<String, OAuthAccessGrant> grantsBySubject =
+                    grantsBySubjectByAffinity.computeIfAbsent(affinityId, a -> new IdentityHashMap<>());
```
Consider hoisting above 2 lines into `supplyGrantsBySubject(realmIndex, affinityId)`.

dknoma committed:
removed accidental empty newline, updated spec dep to 0.29

dknoma committed:
added `supplyGrantsBySubject`
